cli/command/container/opts.go

```diff
@@ -796,19 +792,22 @@ func parseNetworkAttachmentOpt(ep opts.NetworkAttachmentOpts) (*networktypes.End
 	return epConfig, nil
 }
 
-func parsePortOpts(publishOpts []string) ([]string, error) {
+func convertToStandardNotation(ports []string) ([]string, error) {
 	optsList := []string{}
-	for _, publish := range publishOpts {
-		params := map[string]string{"protocol": "tcp"}
-		for _, param := range strings.Split(publish, ",") {
-			opt := strings.Split(param, "=")
-			if len(opt) < 2 {
-				return optsList, errors.Errorf("invalid publish opts format (should be name=value but got '%s')", param)
+	for _, publish := range ports {
+		if strings.Contains(publish, "=") {
+			params := map[string]string{"protocol": "tcp"}
+			for _, param := range strings.Split(publish, ",") {
+				opt := strings.Split(param, "=")
+				if len(opt) < 2 {
+					return optsList, errors.Errorf("invalid publish opts format (should be name=value but got '%s')", param)
+				}
+				params[opt[0]] = opt[1]
 			}
-
-			params[opt[0]] = opt[1]
+			optsList = append(optsList, fmt.Sprintf("%s:%s/%s", params["target"], params["published"], params["protocol"]))
```

Comment:
Note; this needs to be updated once https://github.com/docker/cli/pull/2252 is merged


Also; looking at how it's serialised to a string: 

- host-ip is not in the serialised form. while this looks to be currently not supported in the long form (we should), having a combination of "long" and "short" options could possibly mean that duplicates go unnoticed (I'd have to check). i.e., `-p 0.0.0.0:8080:80` is the equivalent to `-p published=8080,target=80`, but the latter is serialised to `8080:80` (not `0.0.0.0:8080:80`). Then again, we may already have a problem if `-p 0.0.0.0:8080:80` and `-p 127.0.0.1:8080:80` are both set 🤔 
- the long-form `mode=host` is only supported for swarm services (wondering if it should be added as a no-op option for `docker run` 🤔 

Perhaps with some refactoring, we can re-use the `PortOpt` type for both `docker run` and `docker service` (separating validation to produce errors on options that are not supported by either `docker run` or `docker service`); https://github.com/docker/cli/blob/2634562119dd5b4e64d529d1aa5e16f760382d5f/opts/port.go#L29